File: src/org/lice/parse2/ASTAtomicNode.java
package org.lice.parse2;

import org.lice.model.MetaData;
import org.lice.model.Node;
import org.lice.model.SymbolNode;
import org.lice.model.ValueNode;

public class ASTAtomicNode extends ASTNode {
	ASTAtomicNode(MetaData metaData, Token token) {
		super(metaData);
		this.token = token;
	}

	public Token getToken() { return this.token; }
	private Token token;

	@Override
	public Node accept(Sema sema) {
		switch (token.getType()) {
			case BinNumber: return new ValueNode(org.lice.parse2.Parse.toBinInt(token.getStrValue()), getMetaData());
			case OctNumber: return new ValueNode(org.lice.parse2.Parse.toOctInt(token.getStrValue()), getMetaData());
			case HexNumber: return new ValueNode(org.lice.parse2.Parse.toHexInt(token.getStrValue()), getMetaData());
			case DecNumber: return new ValueNode(Integer.parseInt(token.getStrValue()), getMetaData());
			case LongInteger: return new ValueNode(Long.parseLong(token.getStrValue()), getMetaData());
			case BigInt: return new ValueNode(org.lice.parse2.Parse.toBigInt(token.getStrValue()), getMetaData());
			case BigDec: return new ValueNode(org.lice.parse2.Parse.toBigDec(token.getStrValue()), getMetaData());
			case FloatNumber: return new ValueNode(Float.parseFloat(token.getStrValue()), getMetaData());
			case DoubleNumber: return new ValueNode(Double.parseDouble(token.getStrValue()), getMetaData());
			case StringLiteral: return new ValueNode(token.getStrValue(), getMetaData());
			case Identifier: return new SymbolNode(sema.getSymbolList(), token.getStrValue(), getMetaData());
			case LispKwd:
			case EOI:
				assert false;
				break;
		}
		return null;
	}
}

File: src/org/lice/parse2/ASTAtomicNode.kt
package org.lice.parse2

import org.lice.model.*
import org.lice.util.ParseException

class ASTAtomicNode internal constructor(metaData: MetaData, val token: Token) : ASTNode(metaData) {
	override fun accept(sema: Sema) = when (token.type) {
		Token.TokenType.BinNumber -> ValueNode(token.strValue.toBinInt(), metaData)
		Token.TokenType.OctNumber -> ValueNode(token.strValue.toOctInt(), metaData)
		Token.TokenType.HexNumber -> ValueNode(token.strValue.toHexInt(), metaData)
		Token.TokenType.DecNumber -> ValueNode(Integer.parseInt(token.strValue), metaData)
		Token.TokenType.LongInteger -> ValueNode(java.lang.Long.parseLong(token.strValue), metaData)
		Token.TokenType.BigInt -> ValueNode(token.strValue.toBigInt(), metaData)
		Token.TokenType.BigDec -> ValueNode(token.strValue.toBigDec(), metaData)
		Token.TokenType.FloatNumber -> ValueNode(java.lang.Float.parseFloat(token.strValue), metaData)
		Token.TokenType.DoubleNumber -> ValueNode(java.lang.Double.parseDouble(token.strValue), metaData)
		Token.TokenType.StringLiteral -> ValueNode(token.strValue, metaData)
		Token.TokenType.Identifier -> SymbolNode(sema.symbolList, token.strValue, metaData)
		Token.TokenType.LispKwd, Token.TokenType.EOI -> throw ParseException("")
	}
}
